lib/Mojo/Util.pm

```diff
@@ -379,6 +406,27 @@ sub unquote {
   return $str;
 }
 
+sub urandom_token {
+  my $length   = shift || 44;
+  my @alphabet = @{shift || ["A" .. "Z", "a" .. "z", "0" .. "9"]};
+
+  unless (@alphabet >= 2 && @alphabet <= 256) {
+    croak "alphabet size must be between 2 and 256 elements";
+  }
+
+  my $bias_lim = 256 % @alphabet;
+
+  my (@bytes, @token);
+  while (@token < $length) {
+    @bytes = split "", urandom_bytes(64) unless @bytes;
+    my $num = ord(shift @bytes);
+    next if $num < $bias_lim;
+    push @token, $alphabet[$num % @alphabet];
+  }
+
+  return join "", @token;
+}
```

Comment:
I believe that this `urandom_token()` function is correct, however:

* it may be non-obvious to users how the requested length correlates to cryptographic strength ($`= \mathrm{length} \cdot \log_2\left(|\mathrm{alphabet}|\right)`$). As a consenquence, this very pull request also makes some odd (but harmless!) choices, e.g. 160-bit tokens for CSRF protection and ~262-bit tokens as the default app secret.
* the implementation is somewhat complicated
* it "wastes" random bits due to its bias limit, and also because it reads randomness in chunks of 512 bits

All of these could be side-stepped if choosing a design like [Python's `secrets.token_urlsafe(nbytes)` function](https://docs.python.org/3/library/secrets.html#secrets.token_urlsafe):

* no user defined alphabet. The function hardcodes a particular default alphabet.
* this simplifies the problem from "choose random letters" to "encode the bytes"
* the function's `nbytes` parameter directly relates to the cryptographic strength, which is generally more important than the output string length. Python doesn't document the default value, but CPython chooses 32 bytes, which is perfect for typical applications.

Thus, alternative suggestion for this function:

```perl
use MIME::Base64 qw(encode_base64url);

sub _build_urandom_bytes {
  ...
  return sub {
    my $num = shift || 32;  # read 256 bits by default
    ...
  }
}

sub urandom_urlsafe {
  my $nbytes = shift;
  return encode_base64url(urandom_bytes($nbytes));
}
```

Could also add a `urandom_hex()` function that encodes via `unpack "h*"`, e.g. to allow the `_csrf_helper()` to continue outputting a hex-formatted token. But I don't think that's really necessary.

Reply:
This is vibes based, but the de-biasing code here feels "tricky", and I think it's because this code rejects `$num` from the low end instead of the high end of the range.

For example, with an alphabet of 100, you'd reject `$num < 55`... maybe it would be clearer to reject `$num > 200` instead? 

Rejecting the low end means you can never stop ignoring the mod and you have to think about more boundaries you can be off-by-one at while reading.

When rejecting from the high end one can visualise the code as pulling out the `$num`th element from `@alphabet x int(256 / @alphabet)`, that is extending `@alphabet` until you pass a length 255 and then trimming the end off both `$num` and off the elongated to match an even multiple. 

Obviously the math is the same, but having left-over (biasing) elements at the high end for both mod and for the rejection condition might make it more understandable?

Reply:
- `urandom_token` is replaced with `urandom_urlsafe`: As pointed out by @latk and @guest20, the mod reduction and rejection sampling is a bit complex and not neccecary for solving this PR.

   A secondary purpose of`urandom_token` was to provide a "password generator" with custom alphabet support similar to `Session::Token`, which would be useful not really needed for HMAC secrets.
  
   I've replaced it with `urandom_urlsafe` as suggested by @latk, where 256 bits are encoded as url safe base64.
   
   Tests checking mod bias has also been removed as they are no longer neccecary.
  
- `urandom_bytes`: Afaik it should be ok with `sysread` to keep an open file handle across forks, but it now opens `/dev/urandom` on each invocation rather than
   keeping the fh around, as suggested by @Grinnz. This should be less of a problem now that `urandom_urlsafe` encodes 32 bytes directly. 
   
   The Win32::API `SystemFunction036` object is kept using `state`.

Reply:
@stigtsp You don't need de-biasing if your alphabet has 256 characters in it